### ISSUE TITLE
[Feature Request] Stop Transtion in pre-listener

### ISSUE THREAD

**Issue by patrick-mcdougle** [winzou/state-machine#12]
The pre-events currently can't do anything to stop the transition from happening, it'd be cool if the pre-events were more like voters and if any of them returned false, the transition would fail.

Thoughts?

**Comment by patrick-mcdougle** [comment]
I understand that you could argue that you could stop a transition with can, but let's say you do something in the pre-event, like change state with some third-party. In our case it would be sending shipments to a 3rd party API. If this API call fails, we don't want the transition to continue. Putting this on the can event seems wrong to me.

**Comment by hacfi** [comment]
@patrick-mcdougle In general I agree that it could be useful to let a pre listener stop the transition. In your use-case I would consider modeling the api call failure into the process so you can track those events and define what happens if it fails (retries, notifications etc).

**Comment by blazarecki** [comment]
Hi, I implement this feature in https://github.com/winzou/state-machine/pull/15

If you want to stop the transition, in your subscriber/listener oyou can do:

``` php
$event->setRejected(true);
```

**Comment by winzou** [comment]
Hi, what is the use-case exactly?

If a third-party API fails, I guess that stopping the transition won't be enough. You will anyway need to catch an exception, have a certain behavior, etc. So I don't really see the need of stopping propagation in that cas.

Just stopping propagation would mean that your app behaves the same if the API call fails or not, except that the status has been changed or not.

**Comment by blazarecki** [comment]
The use case is the one describe by @patrick-mcdougle.

This PR allow you to check if the transition was apply.

So if the API call fail, he can mark the transition as rejected : `$event->setRejected(true);` 

And then check if the transition fail or not.

```php
if (false === $stateMachine->apply('create')) {
    // The transition was not apply. Do stuff.
}
```

If you want I can throw exception instead of returning `false`. 
It will not break the `StateMachineInterface` https://github.com/winzou/state-machine/blob/master/src/SM/StateMachine/StateMachineInterface.php#L29-L39

What do you think @winzou ?

**Comment by winzou** [comment]
But my point is that if a third-party call fails, an exception is *already* thrown. So I don't see the point of throwing another one from the state machine itself.

**Comment by blazarecki** [comment]
When I read @patrick-mcdougle comment I think he mean something like catching an exception from the API in his code.

First he check if the API is available (some dry run) on a pre_transition listener.

``` php
// Listener on PRE_TRANSITION
try {
    // check the api.
} catch(\Exception $e) {
    $event->setRejected(true);
}
```

Then, If the event is not rejected he can make a real **call** the API, else nothing happen.

Am I right @patrick-mcdougle ?

If I'm not I can close the PR.

**Comment by patrick-mcdougle** [comment]
Yeah something like that. What if the third party doesn't throw exceptions, but rather returns a string error or something? I guess the root reason for me proposing this feature is that I'm against using exceptions for normal logic. That's why I'd prefer a solution like the one presented.

**Comment by winzou** [comment]
I've merged #15, thanks guys!
